Update solvers

martinvonk opened the pull request:
# Short description
Update the solver routine.

# Checklist before PR can be merged:
- [x] closes issue #988
- [x] closes issue #1152
- [x] closes issue #1084
- [x] closes issue #1093 
- [x] closes issue #1018 
- [x] is documented
- [x] code formatted and linted with [ruff](https://docs.astral.sh/ruff/)
- [x] type hints
- [x] tests added or expanded
- [x] remove output for all notebooks with changes

martinvonk committed:
remove all references to "dist" and "uniform"

martinvonk committed:
format

martinvonk committed:
move back dists for emcee stuf

martinvonk committed:
remove leftover uniforms

martinvonk committed:
remove dist from set_parameter

martinvonk committed:
bring back self.initial and self.vary

martinvonk committed:
update emcee notebook

martinvonk committed:
don't initialize again

martinvonk committed:
update notebook

martinvonk committed:
move all the stuff that is only for the least_squares solver

martinvonk committed:
move all the stuff that's only for least_squares

martinvonk committed:
solver cleanup and consistency

martinvonk committed:
Merge branch 'dev2' into solver_initialize

martinvonk committed:
Update assertions for distribution parameter in tests

martinvonk committed:
update type hints

martinvonk committed:
Merge branch 'dev2' into solver_initialize

martinvonk committed:
Merge branch 'dev2' into solver_initialize

martinvonk committed:
fix objfunction partial

martinvonk committed:
use partial for lmfit as well

martinvonk committed:
Merge branch 'dev2' into solver_initialize

martinvonk committed:
Merge branch 'dev2' into solver_initialize

martinvonk committed:
Merge branch 'dev2' into solver_initialize

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
remove initialize

martinvonk committed:
create solver directory

martinvonk committed:
move everything to solver directory

martinvonk committed:
give default name to solver

martinvonk committed:
place for tracksolve

martinvonk committed:
update least_squares

raoulcollenteur commented:
Let's try not use any AI to work on this PR because we really want to keep this as clean and simple as possible.

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
fixes after merging dev

martinvonk committed:
move back objective functions

martinvonk committed:
move tracksolve

martinvonk committed:
move fit_report to solver

martinvonk committed:
update fit_report

martinvonk committed:
add solver parameters

martinvonk committed:
remove nfev as attribute

martinvonk committed:
only parse noise to solve, do it internally

martinvonk committed:
fix for nfev

martinvonk committed:
fix for old attributes

martinvonk committed:
fixes for moved tracksolve

martinvonk committed:
fix for typing

martinvonk committed:
remove parameter dist for now

martinvonk committed:
fixes

martinvonk committed:
super init

martinvonk commented:
I'm done for now with this PR. I think I'm done with most of the refactoring. 

The emceesolver does not work anymore now however... It needs a method for adding the dist column to the model parameters dataframe.

Feel free to work on this from now on. I'll be back in May.

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk commented:
Summary of Changes:
This PR updates the solvers to make them more flexible for future implementations, successfully addressing issue #988.

Key Updates:

dist Column Removed: The Parameters DataFrame schema drops the dist column and removes old references to "dist" and "uniform".

API Change in set_parameter: The set_parameter function no longer accepts a dedicated dist argument. Instead, arbitrary properties are now passed through via **kwargs.

Safe Initialization for MCMC: Added a new initialize() method in EmceeSolve that safeguards models by automatically adding a dist column if it's missing (and logging a warning) to prevent crashes during MCMC runs.

Maintenance & Formatting: Updated relevant example notebooks (e.g., the Emcee notebook) to reflect the changes, and ran code formatting/linting using ruff.

martinvonk committed:
extra docs for base.py

martinvonk committed:
fix for mcmc

martinvonk committed:
move fit_report

martinvonk committed:
updates least_squares

dbrakenhoff committed:
Merge remote-tracking branch 'origin/dev' into solver_initialize

dbrakenhoff committed:
Merge remote-tracking branch 'origin/dev' into solver_initialize

dbrakenhoff commented:
I merged dev to get #1178 into this PR.

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
update doc

martinvonk committed:
ruff

martinvonk committed:
Merge branch 'dev' into solver_initialize

raoulcollenteur committed:
Merge branch 'dev' into solver_initialize

raoulcollenteur committed:
Make sure user provided kwargs are now stored,

 also from upstream solvers and when saving to a dict and thus .pas-file.

raoulcollenteur committed:
Some docstring improvents

raoulcollenteur committed:
Move add_solver up with the other methods, fix bug for ml.fit_report(corr=True)

raoulcollenteur committed:
Restructuring and renaming mcmc solver

raoulcollenteur committed:
Update test_objective_functions.py

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
move solve timers

martinvonk committed:
make nparam a property

raoulcollenteur committed:
Merge branch 'dev' into solver_initialize

raoulcollenteur committed:
rename likelihood function file and fix decorator

raoulcollenteur committed:
add solver parameters to model and enable set_parameter method.

raoulcollenteur committed:
add parameter columns

raoulcollenteur committed:
EmceeSolve works again

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
rename BaseSolver to SolverBase

martinvonk committed:
rename BaseLeastSquares to LeastSquaresBase

martinvonk committed:
make LeastSquaresBase non instantiable

raoulcollenteur commented:
One option I was thinking of for the parameter dataframe and changing columns is to make a ParameterDataFrame class where the solver would register a new column that would than show up in all dataframes. Don't know if this is possible but if so, that might be a plan for the future. 

For now I kinda like the simple approach taken now in `ml.get_init_parameters` where adding the parameters for the solver adds the parameter columns, and setting them through ml.set_parameter even adds it back into the object..

raoulcollenteur committed:
Merge branch 'dev' into solver_initialize

raoulcollenteur committed:
Fix most tests

raoulcollenteur committed:
Fix notebooks

raoulcollenteur committed:
Update test_objective_functions.py

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
create get_stderr method

martinvonk committed:
update to_dict methods

martinvonk committed:
move misfit to seperate file

martinvonk committed:
fix tests

martinvonk committed:
fix test

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
ruff

martinvonk commented:
> One option I was thinking of for the parameter dataframe and changing columns is to make a ParameterDataFrame class where the solver would register a new column that would than show up in all dataframes. Don't know if this is possible but if so, that might be a plan for the future.
> 
> For now I kinda like the simple approach taken now in `ml.get_init_parameters` where adding the parameters for the solver adds the parameter columns, and setting them through ml.set_parameter even adds it back into the object..

I agree that the approach in this PR is good! However, I'm thinking that we might want to remove the stderr column by default as well, and onlly add that if you have a LeastSquares solver...

But I think we should make a new issue on how to handle parameters. If you have for instance 100 realizations with 100 initial parameter sets, then we might want to change things. Good to have a discussion on that in an issue.

martinvonk commented:
> > One option I was thinking of for the parameter dataframe and changing columns is to make a ParameterDataFrame class where the solver would register a new column that would than show up in all dataframes. Don't know if this is possible but if so, that might be a plan for the future.
> > For now I kinda like the simple approach taken now in `ml.get_init_parameters` where adding the parameters for the solver adds the parameter columns, and setting them through ml.set_parameter even adds it back into the object..
> 
> I agree that the approach in this PR is good! However, I'm thinking that we might want to remove the stderr column by default as well, and onlly add that if you have a LeastSquares solver...
> 
> But I think we should make a new issue on how to handle parameters. If you have for instance 100 realizations with 100 initial parameter sets, then we might want to change things. Good to have a discussion on that in an issue.

I opened a discussion: https://github.com/orgs/pastas/discussions/1223

martinvonk committed:
fix typo

martinvonk committed:
fix notebooks

martinvonk committed:
use np.random.default_rng

martinvonk committed:
fix description

raoulcollenteur commented:
- @martinvonk remove stderr everywhere and have LS add it
- @raoulcollenteur fix fit_report for emceesolve.

After this the PR will be opened for review.

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
add stderr column only after solve

martinvonk committed:
clarify message

martinvonk committed:
fix nb

martinvonk committed:
align order of parameter columns

dbrakenhoff committed:
improve verbose series validation
- when verbose, we want to print all errors, and try to avoid raising errors
- shortcut return when series is DataFrame
- add support for failing equidistant check because index is not datetime etc. Instead of raising, print a more vague error message

Improve notebook
- use verbose everywhere and uncomment validate_series calls
- improve suggested solution for dtype not float- add info that pastas supports NaNs to to example notebook

martinvonk committed:
solver result as a bool and dataframe

martinvonk committed:
get rid of _solve_success attribute

dbrakenhoff committed:
delete cell

martinvonk committed:
fix handling of more parametrs in io base model load

martinvonk committed:
fix for fit_constant

martinvonk committed:
drop dist if no emceesolve

martinvonk committed:
.solve_success removed

martinvonk committed:
fix for emcee

martinvonk committed:
fix for models with no solver in the pas file

martinvonk committed:
fix bug for ps.One #1161

martinvonk committed:
remove stderr for emcee

martinvonk committed:
fix emcee models

martinvonk committed:
ruff

martinvonk committed:
compare model with noisemodel

martinvonk committed:
fix for linear trend

martinvonk committed:
fix tests

martinvonk committed:
fix tests

martinvonk committed:
fixes for loading models

martinvonk commented:
I'm done :)

martinvonk committed:
allow users to choose tight_layout

martinvonk committed:
merge results_mosaic in results

martinvonk committed:
fix types and variable names for clarity

martinvonk committed:
fix summary plot

martinvonk committed:
update diagnostics plot to subplot_mosaic as well

martinvonk committed:
some fixes

martinvonk committed:
small bug fix

martinvonk committed:
deprecate results_mosaic

martinvonk committed:
fixes for kwargs and figsize

martinvonk committed:
add back deprecationwarning for split

raoulcollenteur committed:
Apply suggestion from @OnnoEbbens

Co-authored-by: OnnoEbbens <onnoebbens@gmail.com>

raoulcollenteur committed:
Finalize mcmc notebook with new solver setup

raoulcollenteur committed:
Add fit_report for mcmc

raoulcollenteur committed:
Add common fit_report interface

raoulcollenteur committed:
Merge branch 'dev' into solver_initialize

raoulcollenteur committed:
fix the docs

raoulcollenteur committed:
Merge branch 'solver_initialize' of https://github.com/pastas/pastas into solver_initialize

raoulcollenteur committed:
Add solvers to the docs

raoulcollenteur committed:
Add all solvers, objfuncs and likelihoods to the docs

dbrakenhoff committed:
Fixes some things in check module
- only check uncertainty for parameters that vary
- show abs param value in statistic column for uncertainty check
- set name of model as index name for check report dataframe
- support kwargs in stat_ufunc_threshold and rsq_geq_threshold.

dbrakenhoff committed:
Improve warning message
- Check msg reports annual maximum prec as 12 m/d which is confusing, should be just 12 m (or 12m/yr)

raoulcollenteur committed:
Add type hints throughout and fix issues #1152, #1084

raoulcollenteur committed:
Improve model.solve docstring to fix #1018

raoulcollenteur commented:
This PR is ready for review! The actual solutions from ml.solve won't change (i.e., see the notebooks on the docs), but the internal structure is made future-proof to accommodate other solvers better, such as the upcoming PestSolve and MaximumLikelihoodSolve in Pastas > 2.0.

- solvers can now add columns to the ml.parameter dataframe, see for example the emceesolve.
- solvers have a custom fit_report
- There is a solver baseclass
- Documentation is strongly improved

It should be simpler to implement a new solver now, which we tested when I reimplemented mcmc again (it worked well :))

raoulcollenteur commented on pastas/solver/least_squares.py:
```diff
@@ -510,34 +720,98 @@ class LeastSquares(BaseSolver):
         pcov: DataFrame | None = None,
-        nfev: int | None = None,
         **kwargs,
     ) -> None:
```
@martinvonk please click resolve on this.

martinvonk committed:
fix stacked_results

martinvonk committed:
cleaner code

martinvonk committed:
make decomposition less high

dbrakenhoff committed:
comments @OnnoEbbens

dbrakenhoff commented on pastas/io/base.py:
```diff
@@ -135,29 +138,60 @@ def _load_model(data: dict) -> Model:
+    # Merge defaults with file parameters: file values win, defaults fill gaps.
+    file_parameters = data["parameters"].copy()
+
+    # Fix old parameter names for One response functions to match the naming convention from Pastas 2.0
```
Maybe add a note here that this extra logic can be removed when pas-files < v2.0 are no longer supported?

dbrakenhoff commented on pastas/io/base.py:
```diff
@@ -135,29 +138,60 @@ def _load_model(data: dict) -> Model:
+        if solver_key not in data:
+            continue
+        if solver_key == "fit":
+            logger.warning(
```
So are these files still supported now? So this is our warning to say, that these files will potentially no longer work in the future? Should we make that more explicit? If this is just a warning, can't they just load/save with Pastas 2.0?

martinvonk replied:
Good point.

dbrakenhoff commented on pastas/io/base.py:
```diff
@@ -135,29 +138,60 @@ def _load_model(data: dict) -> Model:
-            "The solver object is stored in the model.solver attribute since Pastas "
-            "1.3. Please update your pas-file to the new format by loading and saving "
-            "the file with Pastas 1.3."
+            "The 'stderr' column in the parameters table is longer used if "
```
Why can't users load/save with 2.0?

martinvonk replied:
User can load the 'stderr' column, but only if the solver is `LeastSquares` or `LmfitSolve`. Adding `LeastSquares` or `LmfitSolve` to the model, adds the `stderr` column to the model parameters dataframe. If this column (and solver) is present it, the 'stderr' from the pas file is added to the model parameters dataframe.

dbrakenhoff replied:
So if I understand you correctly, there are pas-files that cannot be loaded with 2.0 when stderr is present in the parameters dataframe, and the solver is not LmfitSolve or LeastSquares. And that is why the warning suggests using 1.5 to update the pas file?

My guess is the number of saved models out there with alternative solvers is very very small? EmceeSolver never supported to_dict, and I'm not sure if some of your other solvers support that? So there is a very high probability you are able to load these older pas-files with 2.0? Just wondering out loud if it's worth the warning, when there is a very high chance the load will succeed with pastas 2.0.

martinvonk replied:
I see the total message better now. The Pastas 1.5 update suggestion should be removed. That has nothing to do with this..

To answer the rest, no the probability is not very high that: there is a not null "stderr" column while the solver is None or EmceeSolve. So I'd be fine with removing this message.

dbrakenhoff commented on pastas/solver/least_squares.py:
```diff
@@ -0,0 +1,1220 @@
+        alpha: float, optional
+            Significance level for the confidence interval. Default is 0.05, which corresponds to a 95% confidence interval.
+        max_iter: int, optional
+            Maximum number of iterations for truncated multivariate sampling, default is 10. Increase this value if number of accepted parameter samples is lower than n.
```
line lengths?

dbrakenhoff commented on pastas/solver/least_squares.py:
```diff
@@ -0,0 +1,1220 @@
+        """
+        Wrapper for the shared `objfunction` to calculate residuals or noise.
+        """
+        return misfit(
```
nitpicky: bit confusing to shadow method name

martinvonk replied:
agree, we don't need it anymore as well.

dbrakenhoff commented on pastas/solver/least_squares.py:
```diff
@@ -0,0 +1,1220 @@
+        if self.method == "lm":
+            logger.info(
+                "Method 'lm' does not support boundaries. Ignoring Pastas'"
+                "`pmin` and `pmax` parameter bounds and setting them to `nan`."
```
~+/-inf actually?~

nvm, this was about pmin/pmax

dbrakenhoff commented on pastas/solver/least_squares.py:
```diff
@@ -0,0 +1,1220 @@
+
+        Notes
+        -----
+        This method is copied from Scipy (version 1.14.1):
```
Separate issue, but should we check if this method ever got any updates in scipy in the meantime?

dbrakenhoff commented on pastas/solver/least_squares.py:
```diff
@@ -0,0 +1,1220 @@
+class LmfitSolve(LeastSquaresBase):
+    """Solving the model using the LmFit :cite:p:`newville_lmfitlmfit-py_2019`.
+
+    This is basically a wrapper around the scipy solvers, adding some cool
```
not very scientific or formal language, but it is cool :)

dbrakenhoff commented on pastas/solver/mcmc.py:
```diff
@@ -0,0 +1,531 @@
+    Parameters
+    ----------
+    objfunction: func, optional
+        An objective function to be minimized. See the pastas.likelihood_functions module for more information.
```
pastas.solver.likelihood? Or just refer explicitly to GaussianLikelihood, etc?

dbrakenhoff replied:
Ah I see it was imported so should be `pastas.likelihood`

dbrakenhoff commented on pastas/solver/mcmc.py:
```diff
@@ -0,0 +1,531 @@
+    emcee.EnsembleSampler
+    emcee.moves
+    emcee.backend
+    pastas.objective_functions
```
pastas.solver.objective_function

dbrakenhoff commented on pastas/solver/mcmc.py:
```diff
@@ -0,0 +1,531 @@
+
+    def __init__(
+        self,
+        name: str = "solver",
```
Is it every useful for solvers to set their name attribute? Seems like it's never used and always set to solver now?

dbrakenhoff commented on pastas/solver/mcmc.py:
```diff
@@ -0,0 +1,531 @@
+        # Set initial positions of the walkers
+        pinit = self.initial[self.vary]
+        ndim = pinit.size
+        pinit = pinit + np.abs(pinit) * 1e-2 * np.random.randn(self.nwalkers, ndim)
```
is this 1e-2 value something you would potentially want to tweak (per parameter)?

dbrakenhoff commented on pastas/solver/mcmc.py:
```diff
@@ -0,0 +1,531 @@
+
+        # Create sampler and run mcmc
+        if self.parallel:
+            logger.info("Going into the parallel universe")
```
🤓

dbrakenhoff commented on pastas/solver/mcmc.py:
```diff
@@ -0,0 +1,531 @@
+            Numpy array with the parameters.
+        noise: bool
+
+        weights
```
docstring somewhat incomplete

dbrakenhoff commented on pastas/solver/trackers.py:
```diff
@@ -0,0 +1,427 @@
+"""Place to put TrackSolve"""
```
Haha, great description

dbrakenhoff commented on pastas/solver/trackers.py:
```diff
@@ -0,0 +1,427 @@
+        ml.oseries.series
+    update_iter : int, optional
+        if visualizing optimization progress, update plot every update_iter
+        iterations, by default nparam
```
missing pause in docstring

dbrakenhoff commented on pastas/model.py:
```diff
@@ -828,21 +820,24 @@ def solve(
-            of Pastas 0.23, not a class!
-        report: bool | Literal["full"], optional
+            Instance of a pastas Solver class used to solve the model. See pastas.
+            solver for the options. If None, the solver from ml.solver is used, and if
```
This is a bit confusing? If you provide nothing you get LeastSquares, when is ml.solver not None? When you're loading an already solved model I guess? Sentence could be clearer.

dbrakenhoff commented on pastas/model.py:
```diff
@@ -828,21 +820,24 @@ def solve(
-            parameters, or False to suppress the report. This can also be
-            manually triggered after optimization by calling print(ml.fit_report(
-            )) on the Pastas model instance.
+            parameters, or False to suppress the report. To have full conrol over the
```
Also a bit confusing? What is the extra control I can have, and why would I want that? Maybe just mention the other methods without too much extra and maybe unnecessary info?

dbrakenhoff commented on pastas/model.py:
```diff
@@ -874,9 +869,14 @@ def solve(
         reset_settings: bool = False,
-            If True, the model settings are reset to their default values before solving.
-            This calls the Model.set_settings() method with default values.
+            If True, the model settings are reset to their default values before
```
So initial is deprecated in favor of reset_settings. reset_settings mentions returning all model settings back to their defaults. It does not mention anything about initial parameters. Should we add that?

Also the name reset_settings to me indicates we're resetting the settings, which could in theory be separate from the initial parameter values? So this isn't immediately clear to me from reading this.

dbrakenhoff commented on pastas/model.py:
```diff
@@ -977,34 +976,34 @@ def solve(
-            logger.warning("Model parameters could not be estimated well.")
+        solve_success, result = self.solver.solve(weights=weights, **kwargs)
+
+        # Update the parameters with the results from the optimization
```
These parameters do not have to be set using ml.set_parameter() because they only live in the model class, right? E.g. optimal, stderr, etc.

dbrakenhoff commented on pastas/model.py:
```diff
@@ -1138,11 +1137,13 @@ def set_parameter(
         move_bounds: bool, optional
-            Reset pmin/pmax based on new initial value. Of move_bounds=True, pmin and
-            pmax must be None.
+            Reset pmin/pmax based on new initial value. Of move_bounds=True,
```
Of --> If

dbrakenhoff commented on pastas/model.py:
```diff
@@ -2036,7 +2061,7 @@ def _generate_warnings_report(self, log: bool = True) -> list[str]:
         # model optimization unsuccessful
-        if not self._solve_success:
+        if solve_success is False:
             msg.append("Model parameters could not be estimated well.")
```
can we provide any guidance with this message where to find more diagnostics?

dbrakenhoff reviewed (approved):
Really nice work. Much cleaner, and ready for future developments! 

I commented on a few minor docstring things! But approving in advance!

martinvonk committed:
Refactor noise model handling and update deprecation warnings for legacy pas-files

martinvonk committed:
Remove deprecated misfit method from LeastSquaresBase class

martinvonk committed:
fix line lenghts

martinvonk committed:
Refactor fit_report method parameters to use 'full_output' instead of 'all_options' for consistency with baseclass

martinvonk committed:
formalize language

martinvonk committed:
Update references to SciPy source in LeastSquares class documentation for accuracy

martinvonk committed:
Update docstring for TrackSolve class to clarify callback methods and add pause parameter

martinvonk committed:
Clarify objfunction parameter type in EmceeSolve class documentation and update import path for objective functions

martinvonk committed:
fix typo

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk commented:
Thanks for the review @dbrakenhoff. I fixed some of your comments and replied to some other. Some comments are still open for @raoulcollenteur :)

I also made a small change to the fit_report @raoulcollenteur made because it was incompatible with the SolverBase class. Also I changed the name `all_options` to `full_output` because I thought `all_options` was a but vague.. Maybe `full_report` would be even better as a variable name..

Do you also have some time to look at this PR @OnnoEbbens? I thought you had some things you wanted to test with the new solver?

dbrakenhoff commented:
> I also made a small change to the fit_report @raoulcollenteur made because it was incompatible with the SolverBase class. Also I changed the name `all_options` to `full_output` because I thought `all_options` was a but vague.. Maybe `full_report` would be even better as a variable name..

I think `full_output=True` is nice, similar to what scipy uses in some of their optimization funcs. If you're considering `fit_report(full_report=True)` I think calling `fit_report(full=True)` is also pretty clear to me.

martinvonk committed:
fix for misfit movement

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
Merge branch 'dev' into solver_initialize

martinvonk committed:
merge threshold base.py

martinvonk committed:
fix for p larger than rfunc.nparam

martinvonk committed:
ruff

raoulcollenteur committed:
couple of docstring changes + error for ps.io.load on 1.3

martinvonk committed:
rename emceeSolve to Emcee

martinvonk committed:
rename LmfitSolve to Lmit

martinvonk committed:
deprecate emceesolve and lmfitsolve in pastas namespace

martinvonk committed:
rename solvers in examples

martinvonk committed:
remove LeastSquares from pastas namespace

raoulcollenteur commented:
@martinvonk looks like you made the required changes already and this can be merged after or so I miss something?

martinvonk committed:
fix for solver_name in base.py

martinvonk committed:
Merge branch 'dev' into solver_initialize

dbrakenhoff committed:
Merge branch 'dev' into solver_initialize

dbrakenhoff commented on pastas/solver/least_squares.py:
```diff
@@ -1069,7 +1069,15 @@ def to_dict(self) -> dict:
 
-class LmfitSolve(LeastSquaresBase):
+@PastasDeprecationWarning(
+    version="2.3.0", reason="The EmceeSolve class is renamed to Emcee."
```
wrong message?

martinvonk committed:
fix messages

martinvonk committed:
fix names

martinvonk committed:
remove logger warning
